Refactor bulk move controller

NobodysNightmare opened the pull request:
## OpenProject work packages
- Refactoring: https://community.openproject.org/work_packages/16342
- Also fixed a bug:
  - https://community.openproject.org/work_packages/5798 (public)
  - https://community.openproject.org/work_packages/20101 (private)
## Description

This PR rebases the work of https://github.com/opf/openproject/pull/1974 onto the current dev and applies further refactorings and fixes.
## Notes

:warning: **whoever merges this:** :warning: 

This PR applies cleanly to `release/4.1`. I opened the PR against `dev` to be safe, but we can still decide to merge it into `release/4.1` first. At the moment my opinion is, that it would be worth it.

slawa85 committed:
Common functionality from move and bulk edit work package controller

slawa85 committed:
Move functionality to work package service

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-refactor-bulk-move-controller.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-refactor-bulk-move-controller.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
fix styling of bulk & move stuff

previously destroyed when rebasing a PR onto the dev

NobodysNightmare commented on app/services/work_package_bulk_update_service.rb:
```diff
@@ -0,0 +1,157 @@
+
+  def prepare(params)
+    @work_packages.sort!
+    @copy             = params.has_key? :copy
```
Repeating my comment from the original PR:

Feedback from my side is coming late now, but this copy-parameter looks kind of messy.

The method `save` has a parameter called `params` and one called `copy`, the former passed into this method and checking whether it contains a key called `:copy`.

This is the first red flag for me, because the `params` hash may contain a key called `copy` as well as there may be the copy parameter.

The second red flag is that it suffices that the `:copy` key is present for `@copy` to be true. That means `{ copy: false }` would make `@copy` true.

NobodysNightmare commented on app/controllers/work_packages/bulk_controller.rb:
```diff
@@ -39,34 +39,18 @@ class WorkPackages::BulkController < ApplicationController
+    WorkPackageBulkUpdateService.new(@work_packages, @projects).run(params)
+      .available_attributes.each do |key, value|
+        instance_variable_set(:"@#{key}", value)
+      end
```
I don't see why calling `instance_variable_set` in a loop is a better idea than explicitly setting three instance variables by hand.

NobodysNightmare commented on app/controllers/work_packages/moves_controller.rb:
```diff
@@ -33,46 +33,24 @@ class WorkPackages::MovesController < ApplicationController
+    WorkPackageBulkUpdateService.new(@work_packages, @projects)
+      .run(params).available_attributes.each do |key, value|
+        instance_variable_set(:"@#{key}", value)
+      end
```
I don't see why calling `instance_variable_set` in a loop is a better idea than explicitly setting three instance variables by hand.

NobodysNightmare commented:
Another problem:

The service often writes available values that are suitable when called from the bulk controller, but simply wrong when called from the move controller. This raises no errors, because the corresponding views selectively ignore the values not meant for them.

TBH: I like the idea of refactoring out a single service for both controllers and I start to see what both controllers need. But the current way simply adds another class to the picture with spaghetti-like dependencies without any quality gains.

NobodysNightmare committed:
unify usage of available attributes

- service generates available attributes usable by both controllers
- moves and bulk views both use available attributes provided by service more consequently

NobodysNightmare committed:
clarify some variable names

NobodysNightmare committed:
return wp lists for moved and not moved WPs

not ids for one kind and wps for the other kind

Commit:
Adapt version to corresponding OpenProject core version.

NobodysNightmare committed:
actually allow the status to be set while moving

apparently the status_id param was set to expect an array of values while being moved in this commit: d7f399ab8db059153d2792f9db94450150c7ffdb
Probably because the spec started to expect that when it was moved to be an RSpec test.
I am not sure if this was ever a valid thing outside the tests...

NobodysNightmare committed:
style

NobodysNightmare commented:
https://community.openproject.org/work_packages/5798 
    https://community.openproject.org/work_packages/20101

NobodysNightmare committed:
unify method naming

NobodysNightmare committed:
replace ternary operator by condition
